janusgraph-core/src/main/java/org/janusgraph/graphdb/query/index/IndexSelectionUtil.java

```diff
@@ -41,22 +42,57 @@ public static boolean indexCoversOrder(MixedIndexType index, OrderList orders) {
         return true;
     }
 
+    private static Iterable<IndexType> getKeyIndexesForCondition(Condition<JanusGraphElement> condition) {
+        if (condition instanceof PredicateCondition) {
+            final RelationType type = ((PredicateCondition<RelationType, JanusGraphElement>) condition).getKey();
+            Preconditions.checkArgument(type != null && type.isPropertyKey());
+            return ((InternalRelationType) type).getKeyIndexes();
+        }
+        return Collections.emptySet();
+    }
+
     public static Set<IndexType> getMatchingIndexes(MultiCondition<JanusGraphElement> conditions) {
-        if (conditions == null) {
+        return getMatchingIndexes(conditions, i -> true);
+    }
+
+    public static Set<IndexType> getMatchingIndexes(MultiCondition<JanusGraphElement> conditions, Predicate<IndexType> filter) {
+        if (conditions == null || filter == null) {
             return Collections.emptySet();
         }
         final Set<IndexType> availableIndexes = new HashSet<>();
         ConditionUtil.traversal(conditions, condition -> {
-            if (condition instanceof PredicateCondition) {
-                final RelationType type = ((PredicateCondition<RelationType, JanusGraphElement>) condition).getKey();
-                Preconditions.checkArgument(type != null && type.isPropertyKey());
-                Iterables.addAll(availableIndexes, ((InternalRelationType) type).getKeyIndexes());
+            for (IndexType indexCandidate : getKeyIndexesForCondition(condition)) {
+                if (filter.test(indexCandidate)) {
+                    availableIndexes.add(indexCandidate);
+                }
             }
             return true;
         });
         return availableIndexes;
     }
 
+    public static boolean existsMatchingIndex(MultiCondition<JanusGraphElement> conditions)  {
+        return existsMatchingIndex(conditions, i -> true);
+    }
+
+    public static boolean existsMatchingIndex(MultiCondition<JanusGraphElement> conditions, Predicate<IndexType> filter) {
+        if (conditions == null || filter == null) {
+            return false;
+        }
+
+        Set<Condition<JanusGraphElement>> conditionSet = new HashSet<>();
+        ConditionUtil.traversal(conditions, conditionSet::add);
```

Comment:
I'm not quite happy with this way of dealing with `MultiCondition`s. But on first glance, the simplest way of iterating `Condition`s is to aggregate them in a `Set` first.

Reply:
I think we can omit processing unnecessary conditions and collect them into a set by using apache lang MutableBoolean instead. Something like:
```
public static boolean existsMatchingIndex(MultiCondition<JanusGraphElement> conditions, Predicate<IndexType> filter) {
    if (conditions == null || filter == null) {
        return false;
    }
    MutableBoolean exists = new MutableBoolean();
    ConditionUtil.traversal(conditions, condition -> {
        if(exists.isFalse()){
            for (IndexType indexCandidate : getKeyIndexesForCondition(condition)) {
                if (filter.test(indexCandidate)) {
                    exists.setTrue();
                    return true;
                }
            }
        }
        return true;
    });
    return exists.booleanValue();
}
```

Reply:
I thought about that, but the `exists.isFalse()` condition didn't come to my mind. Like!